File: packages/agent/src/infrastructure/events/types/message.ts
/**
 * @fileoverview Message Events
 *
 * Events for user, assistant, and system messages.
 */

import type { BaseEvent } from './base.js';
import type { TokenUsage } from './token-usage.js';
import type { NormalizedTokenUsage } from './streaming.js';

// =============================================================================
// Content Block Types
// =============================================================================

/** Content block types for messages */
export type ContentBlock =
  | { type: 'text'; text: string }
  | { type: 'image'; source: { type: 'base64'; mediaType: string; data: string } }
  | { type: 'tool_use'; id: string; name: string; input: Record<string, unknown> }
  | { type: 'tool_result'; toolUseId: string; content: string; isError?: boolean }
  | { type: 'thinking'; thinking: string };

// =============================================================================
// Message Events
// =============================================================================

/**
 * User message event
 */
export interface UserMessageEvent extends BaseEvent {
  type: 'message.user';
  payload: {
    content: string | ContentBlock[];
    /** Turn number within session */
    turn: number;
    /** Optional attached images */
    imageCount?: number;
  };
}

/**
 * Assistant message event
 */
export interface AssistantMessageEvent extends BaseEvent {
  type: 'message.assistant';
  payload: {
    content: ContentBlock[];
    turn: number;
    tokenUsage: TokenUsage;
    /**
     * Normalized token usage with semantic clarity for different UI components.
     * Handles provider semantic differences (Anthropic vs OpenAI/Codex/Gemini).
     * This is stored directly on message.assistant so iOS can reconstruct without
     * correlating with stream.turn_end events.
     */
    normalizedUsage?: NormalizedTokenUsage;
    stopReason: 'end_turn' | 'tool_use' | 'max_tokens' | 'stop_sequence';
    /** Duration of LLM call in ms */
    latency?: number;
    /** Model used (may differ from session default) */
    model: string;
    /** Whether extended thinking was used */
    hasThinking?: boolean;
  };
}

/**
 * System message event
 */
export interface SystemMessageEvent extends BaseEvent {
  type: 'message.system';
  payload: {
    content: string;
    source: 'compaction' | 'context' | 'hook' | 'error' | 'inject';
  };
}

File: packages/agent/crates/tron-events/src/types/payloads/message.rs
//! Message event payloads: user, assistant, system.

use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::token_usage::{TokenRecord, TokenUsage};

/// Payload for `message.user` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UserMessagePayload {
    /// User message content — either a plain string or array of content blocks.
    pub content: Value,
    /// Turn number.
    pub turn: i64,
    /// Number of images attached.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub image_count: Option<i64>,
    /// Skills attached to this message (raw iOS objects: `[{name, source, displayName}]`).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skills: Option<Value>,
    /// Spells attached to this message (same format as skills).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub spells: Option<Value>,
}

/// Payload for `message.assistant` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AssistantMessagePayload {
    /// Content blocks (text, `tool_use`, thinking).
    pub content: Value,
    /// Turn number.
    pub turn: i64,
    /// Token usage for this message.
    pub token_usage: TokenUsage,
    /// Canonical token record.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub token_record: Option<TokenRecord>,
    /// LLM stop reason.
    pub stop_reason: String,
    /// LLM call latency in ms.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub latency: Option<i64>,
    /// Model ID used.
    pub model: String,
    /// Whether the response included thinking blocks.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub has_thinking: Option<bool>,
}

/// Payload for `message.system` events.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SystemMessagePayload {
    /// System message content.
    pub content: String,
    /// Source of the system message.
    pub source: String,
}
